File: .skills/openclaw-skills/skills/0xnyk/xint/lib/moderation.ts
/**
 * moderation.ts — OAuth-authenticated block/mute management.
 * Supports list/add/remove for both blocked and muted users.
 */

import { BASE, oauthDelete, oauthGet, oauthPost } from "./api";
import { getValidToken, loadTokens } from "./oauth";
import { trackCost } from "./costs";

interface XUser {
  id: string;
  username: string;
  name?: string;
  description?: string;
  public_metrics?: {
    followers_count?: number;
  };
}

interface ModeConfig {
  noun: "blocks" | "mutes";
  path: "blocking" | "muting";
  listOp: "blocks_list" | "mutes_list";
  addOp: "blocks_add" | "mutes_add";
  removeOp: "blocks_remove" | "mutes_remove";
}

const USER_FIELDS = "user.fields=id,username,name,description,public_metrics";

export function isLikelyUserId(value: string): boolean {
  return /^\d+$/.test(value);
}

export function parseListFlags(args: string[]): { json: boolean; limit: number } {
  let json = false;
  let limit = 50;

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "--json":
        json = true;
        break;
      case "--limit": {
        const parsed = parseInt(args[++i] || "50", 10);
        limit = Math.max(1, Number.isNaN(parsed) ? 50 : parsed);
        break;
      }
    }
  }

  return { json, limit };
}

async function resolveUserId(input: string, accessToken: string): Promise<{ id: string; username: string }> {
  const value = input.replace(/^@/, "");
  if (isLikelyUserId(value)) {
    return { id: value, username: value };
  }

  const url = `${BASE}/users/by/username/${encodeURIComponent(value)}?${USER_FIELDS}`;
  const raw = await oauthGet(url, accessToken);
  const user = raw?.data;
  if (!user?.id) {
    throw new Error(`User not found: @${value}`);
  }
  return { id: user.id, username: user.username || value };
}

async function fetchUsers(
  userId: string,
  accessToken: string,
  cfg: ModeConfig,
  maxTotal: number,
): Promise<XUser[]> {
  const all: XUser[] = [];
  let nextToken: string | undefined;

  while (all.length < maxTotal) {
    const perPage = Math.min(100, maxTotal - all.length);
    let url = `${BASE}/users/${userId}/${cfg.path}?max_results=${perPage}&${USER_FIELDS}`;
    if (nextToken) url += `&pagination_token=${nextToken}`;

    const raw = await oauthGet(url, accessToken);
    const users: XUser[] = raw?.data || [];
    if (users.length === 0) break;

    all.push(...users);
    nextToken = raw?.meta?.next_token;
    if (!nextToken) break;
  }

  return all.slice(0, maxTotal);
}

async function addUser(userId: string, targetUserId: string, accessToken: string, cfg: ModeConfig): Promise<boolean> {
  const url = `${BASE}/users/${userId}/${cfg.path}`;
  const result = await oauthPost(url, accessToken, { target_user_id: targetUserId });
  return result?.data?.blocking === true || result?.data?.muting === true || result?.success === true;
}

async function removeUser(
  sourceUserId: string,
  targetUserId: string,
  accessToken: string,
  cfg: ModeConfig,
): Promise<boolean> {
  const url = `${BASE}/users/${sourceUserId}/${cfg.path}/${targetUserId}`;
  const result = await oauthDelete(url, accessToken);
  return result?.data?.blocking === false || result?.data?.muting === false || result?.success === true;
}

function usage(mode: "blocks" | "mutes"): void {
  console.log(`Usage: xint ${mode} <subcommand> [options]

Subcommands:
  list [--limit N] [--json]                 List your ${mode} (default)
  add <@username|user_id> [--json]          Add ${mode === "blocks" ? "block" : "mute"}
  remove <@username|user_id> [--json]       Remove ${mode === "blocks" ? "block" : "mute"}

Examples:
  xint ${mode}
  xint ${mode} add @example
  xint ${mode} remove 2244994945
  xint ${mode} list --limit 100 --json`);
}

async function runMode(mode: "blocks" | "mutes", rawArgs: string[]): Promise<void> {
  const cfg: ModeConfig =
    mode === "blocks"
      ? {
          noun: "blocks",
          path: "blocking",
          listOp: "blocks_list",
          addOp: "blocks_add",
          removeOp: "blocks_remove",
        }
      : {
          noun: "mutes",
          path: "muting",
          listOp: "mutes_list",
          addOp: "mutes_add",
          removeOp: "mutes_remove",
        };

  const sub = (rawArgs[0] || "list").toLowerCase();
  const args = rawArgs.slice(1);

  if (sub === "help" || sub === "--help" || sub === "-h") {
    usage(mode);
    return;
  }

  const tokens = loadTokens();
  if (!tokens) throw new Error("Not authenticated. Run 'auth setup' first.");
  const accessToken = await getValidToken();

  if (sub === "list" || sub === "ls") {
    const { json, limit } = parseListFlags(args);
    const users = await fetchUsers(tokens.user_id, accessToken, cfg, limit);
    trackCost(cfg.listOp, `/2/users/${tokens.user_id}/${cfg.path}`, users.length);

    if (json) {
      console.log(JSON.stringify(users, null, 2));
      return;
    }

    if (users.length === 0) {
      console.log(`No ${cfg.noun} found.`);
      return;
    }

    console.log(`\n🚫 ${cfg.noun[0].toUpperCase()}${cfg.noun.slice(1)} — @${tokens.username} (${users.length})\n`);
    for (let i = 0; i < users.length; i++) {
      const u = users[i];
      const followers = u.public_metrics?.followers_count;
      const followerStr = followers !== undefined ? ` (${followers} followers)` : "";
      console.log(`${i + 1}. @${u.username}${u.name ? ` — ${u.name}` : ""}${followerStr}`);
      if (u.description) {
        console.log(`   ${u.description.slice(0, 180).replace(/\n/g, " ")}`);
      }
    }
    return;
  }

  if (sub === "add") {
    const target = args.find((a) => !a.startsWith("-"));
    const json = args.includes("--json");
    if (!target) throw new Error(`Usage: xint ${mode} add <@username|user_id>`);

    const resolved = await resolveUserId(target, accessToken);
    const ok = await addUser(tokens.user_id, resolved.id, accessToken, cfg);
    trackCost(cfg.addOp, `/2/users/${tokens.user_id}/${cfg.path}`, 0);

    if (json) {
      console.log(
        JSON.stringify({ success: ok, user_id: resolved.id, username: resolved.username, action: `${mode}_add` }, null, 2),
      );
      return;
    }

    console.log(
      ok
        ? `✅ ${mode === "blocks" ? "Blocked" : "Muted"} @${resolved.username}`
        : `Failed to ${mode === "blocks" ? "block" : "mute"} @${resolved.username}`,
    );
    return;
  }

  if (sub === "remove" || sub === "rm" || sub === "delete") {
    const target = args.find((a) => !a.startsWith("-"));
    const json = args.includes("--json");
    if (!target) throw new Error(`Usage: xint ${mode} remove <@username|user_id>`);

    const resolved = await resolveUserId(target, accessToken);
    const ok = await removeUser(tokens.user_id, resolved.id, accessToken, cfg);
    trackCost(cfg.removeOp, `/2/users/${tokens.user_id}/${cfg.path}/${resolved.id}`, 0);

    if (json) {
      console.log(
        JSON.stringify({ success: ok, user_id: resolved.id, username: resolved.username, action: `${mode}_remove` }, null, 2),
      );
      return;
    }

    console.log(
      ok
        ? `✅ Removed ${mode === "blocks" ? "block" : "mute"} for @${resolved.username}`
        : `Failed to remove ${mode === "blocks" ? "block" : "mute"} for @${resolved.username}`,
    );
    return;
  }

  usage(mode);
}

export async function cmdBlocks(rawArgs: string[]): Promise<void> {
  await runMode("blocks", rawArgs);
}

export async function cmdMutes(rawArgs: string[]): Promise<void> {
  await runMode("mutes", rawArgs);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/moderation.rs
use anyhow::Result;

use crate::auth::oauth;
use crate::cli::ModerationArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::format;

const USER_FIELDS: &str = "user.fields=id,username,name,description,public_metrics";

enum Mode {
    Blocks,
    Mutes,
}

impl Mode {
    fn noun(&self) -> &'static str {
        match self {
            Mode::Blocks => "blocks",
            Mode::Mutes => "mutes",
        }
    }

    fn path(&self) -> &'static str {
        match self {
            Mode::Blocks => "blocking",
            Mode::Mutes => "muting",
        }
    }

    fn list_op(&self) -> &'static str {
        match self {
            Mode::Blocks => "blocks_list",
            Mode::Mutes => "mutes_list",
        }
    }

    fn add_op(&self) -> &'static str {
        match self {
            Mode::Blocks => "blocks_add",
            Mode::Mutes => "mutes_add",
        }
    }

    fn remove_op(&self) -> &'static str {
        match self {
            Mode::Blocks => "blocks_remove",
            Mode::Mutes => "mutes_remove",
        }
    }

    fn action_add(&self) -> &'static str {
        match self {
            Mode::Blocks => "block",
            Mode::Mutes => "mute",
        }
    }

    fn action_remove(&self) -> &'static str {
        match self {
            Mode::Blocks => "unblock",
            Mode::Mutes => "unmute",
        }
    }
}

pub async fn run_blocks(args: &ModerationArgs, config: &Config, client: &XClient) -> Result<()> {
    run_mode(Mode::Blocks, args, config, client).await
}

pub async fn run_mutes(args: &ModerationArgs, config: &Config, client: &XClient) -> Result<()> {
    run_mode(Mode::Mutes, args, config, client).await
}

async fn run_mode(
    mode: Mode,
    args: &ModerationArgs,
    config: &Config,
    client: &XClient,
) -> Result<()> {
    let client_id = config.require_client_id()?;
    let (access_token, tokens) =
        oauth::get_valid_token(client, &config.tokens_path(), client_id).await?;

    let parts: Vec<String> = args.subcommand.clone().unwrap_or_default();
    let sub = parts.first().map(String::as_str).unwrap_or("list");

    match sub {
        "list" | "ls" => {
            cmd_list(
                &mode,
                args,
                &tokens.user_id,
                &tokens.username,
                &access_token,
                config,
                client,
            )
            .await
        }
        "add" => {
            cmd_add(
                &mode,
                args,
                &parts,
                &tokens.user_id,
                &access_token,
                config,
                client,
            )
            .await
        }
        "remove" | "rm" | "delete" => {
            cmd_remove(
                &mode,
                args,
                &parts,
                &tokens.user_id,
                &access_token,
                config,
                client,
            )
            .await
        }
        "help" | "--help" | "-h" => {
            print_help(&mode);
            Ok(())
        }
        _ => {
            eprintln!("Unknown {} subcommand: {}", mode.noun(), sub);
            print_help(&mode);
            Ok(())
        }
    }
}

async fn cmd_list(
    mode: &Mode,
    args: &ModerationArgs,
    user_id: &str,
    username: &str,
    access_token: &str,
    config: &Config,
    client: &XClient,
) -> Result<()> {
    let users = fetch_users(
        client,
        user_id,
        mode.path(),
        access_token,
        args.limit.max(1),
    )
    .await?;
    costs::track_cost(
        &config.costs_path(),
        mode.list_op(),
        &format!("/2/users/{}/{}", user_id, mode.path()),
        users.len() as u64,
    );

    if args.json {
        format::print_json_pretty_filtered(&users)?;
        return Ok(());
    }

    if users.is_empty() {
        println!("No {} found.", mode.noun());
        return Ok(());
    }

    println!(
        "\n{} — @{} ({})\n",
        title_case(mode.noun()),
        username,
        users.len()
    );
    for (i, user) in users.iter().enumerate() {
        let handle = user.get("username").and_then(|v| v.as_str()).unwrap_or("?");
        let name = user.get("name").and_then(|v| v.as_str()).unwrap_or("");
        println!(
            "{}. @{}{}",
            i + 1,
            handle,
            if name.is_empty() {
                String::new()
            } else {
                format!(" — {name}")
            }
        );
    }

    Ok(())
}

async fn cmd_add(
    mode: &Mode,
    args: &ModerationArgs,
    parts: &[String],
    source_user_id: &str,
    access_token: &str,
    config: &Config,
    client: &XClient,
) -> Result<()> {
    let target = parts
        .get(1)
        .map(String::as_str)
        .ok_or_else(|| anyhow::anyhow!("Usage: xint {} add <@username|user_id>", mode.noun()))?;

    let (target_user_id, target_username) = resolve_user_id(client, access_token, target).await?;
    let body = serde_json::json!({ "target_user_id": target_user_id });
    let result = client
        .oauth_post(
            &format!("users/{}/{}", source_user_id, mode.path()),
            access_token,
            Some(&body),
        )
        .await?;

    costs::track_cost(
        &config.costs_path(),
        mode.add_op(),
        &format!("/2/users/{}/{}", source_user_id, mode.path()),
        0,
    );

    if args.json {
        format::print_json_pretty_filtered(&result)?;
    } else {
        let success = is_added(&result, mode) || result.get("success").is_some();
        if success {
            println!("{} @{}", past_tense(mode.action_add()), target_username);
        } else {
            println!("Failed to {} @{}", mode.action_add(), target_username);
        }
    }

    Ok(())
}

async fn cmd_remove(
    mode: &Mode,
    args: &ModerationArgs,
    parts: &[String],
    source_user_id: &str,
    access_token: &str,
    config: &Config,
    client: &XClient,
) -> Result<()> {
    let target = parts
        .get(1)
        .map(String::as_str)
        .ok_or_else(|| anyhow::anyhow!("Usage: xint {} remove <@username|user_id>", mode.noun()))?;

    let (target_user_id, target_username) = resolve_user_id(client, access_token, target).await?;
    let result = client
        .oauth_delete(
            &format!(
                "users/{}/{}/{}",
                source_user_id,
                mode.path(),
                target_user_id
            ),
            access_token,
        )
        .await?;

    costs::track_cost(
        &config.costs_path(),
        mode.remove_op(),
        &format!(
            "/2/users/{}/{}/{}",
            source_user_id,
            mode.path(),
            target_user_id
        ),
        0,
    );

    if args.json {
        format::print_json_pretty_filtered(&result)?;
    } else {
        let success = is_removed(&result, mode) || result.get("success").is_some();
        if success {
            println!("{} @{}", past_tense(mode.action_remove()), target_username);
        } else {
            println!("Failed to {} @{}", mode.action_remove(), target_username);
        }
    }

    Ok(())
}

async fn fetch_users(
    client: &XClient,
    user_id: &str,
    path_segment: &str,
    access_token: &str,
    max_total: usize,
) -> Result<Vec<serde_json::Value>> {
    let mut all = Vec::new();
    let mut next_token: Option<String> = None;

    while all.len() < max_total {
        let per_page = (max_total - all.len()).min(100);
        let pagination = match &next_token {
            Some(token) => format!("&pagination_token={token}"),
            None => String::new(),
        };
        let path = format!(
            "users/{user_id}/{path_segment}?max_results={per_page}&{USER_FIELDS}{pagination}"
        );

        let raw = client.oauth_get(&path, access_token).await?;
        let batch = raw
            .data
            .as_ref()
            .and_then(|d| d.as_array())
            .cloned()
            .unwrap_or_default();
        if batch.is_empty() {
            break;
        }
        all.extend(batch);
        next_token = raw.meta.and_then(|m| m.next_token);
        if next_token.is_none() {
            break;
        }
    }

    all.truncate(max_total);
    Ok(all)
}

async fn resolve_user_id(
    client: &XClient,
    access_token: &str,
    input: &str,
) -> Result<(String, String)> {
    let value = input.trim_start_matches('@');
    if is_likely_user_id(value) {
        return Ok((value.to_string(), value.to_string()));
    }

    let path = format!("users/by/username/{value}?{USER_FIELDS}");
    let raw = client.oauth_get(&path, access_token).await?;
    let user = raw
        .data
        .as_ref()
        .ok_or_else(|| anyhow::anyhow!("User @{value} not found"))?;
    let id = user
        .get("id")
        .and_then(|v| v.as_str())
        .ok_or_else(|| anyhow::anyhow!("User @{value} not found"))?;
    let username = user
        .get("username")
        .and_then(|v| v.as_str())
        .unwrap_or(value);
    Ok((id.to_string(), username.to_string()))
}

fn is_added(result: &serde_json::Value, mode: &Mode) -> bool {
    match mode {
        Mode::Blocks => result.pointer("/data/blocking") == Some(&serde_json::Value::Bool(true)),
        Mode::Mutes => result.pointer("/data/muting") == Some(&serde_json::Value::Bool(true)),
    }
}

fn is_removed(result: &serde_json::Value, mode: &Mode) -> bool {
    match mode {
        Mode::Blocks => result.pointer("/data/blocking") == Some(&serde_json::Value::Bool(false)),
        Mode::Mutes => result.pointer("/data/muting") == Some(&serde_json::Value::Bool(false)),
    }
}

fn is_likely_user_id(input: &str) -> bool {
    !input.is_empty() && input.chars().all(|c| c.is_ascii_digit())
}

fn capitalize(s: &str) -> String {
    let mut chars = s.chars();
    match chars.next() {
        Some(first) => first.to_ascii_uppercase().to_string() + chars.as_str(),
        None => String::new(),
    }
}

fn title_case(s: &str) -> String {
    capitalize(s)
}

fn past_tense(action: &str) -> &'static str {
    match action {
        "block" => "Blocked",
        "mute" => "Muted",
        "unblock" => "Unblocked",
        "unmute" => "Unmuted",
        _ => "Updated",
    }
}

fn print_help(mode: &Mode) {
    println!("Usage: xint {} <subcommand> [options]", mode.noun());
    println!();
    println!("Subcommands:");
    println!(
        "  list [--limit N] [--json]                 List your {} (default)",
        mode.noun()
    );
    println!(
        "  add <@username|user_id> [--json]          {} a user",
        capitalize(mode.action_add())
    );
    println!(
        "  remove <@username|user_id> [--json]       {} a user",
        capitalize(mode.action_remove())
    );
}

#[cfg(test)]
mod tests {
    use super::{capitalize, is_likely_user_id, past_tense};

    #[test]
    fn test_is_likely_user_id() {
        assert!(is_likely_user_id("2244994945"));
        assert!(!is_likely_user_id("jack"));
        assert!(!is_likely_user_id("@jack"));
    }

    #[test]
    fn test_capitalize() {
        assert_eq!(capitalize("block"), "Block");
        assert_eq!(capitalize(""), "");
    }

    #[test]
    fn test_past_tense() {
        assert_eq!(past_tense("mute"), "Muted");
        assert_eq!(past_tense("unmute"), "Unmuted");
    }
}
